File: nbomber/LoadTest.cs
using NBomber.CSharp;
using NBomber.Http.CSharp;

namespace NBomberTests;

public static class LoadTest
{
    private const string BaseUrl = "https://jsonplaceholder.typicode.com";

    public static void Run()
    {
        using var httpClient = new HttpClient();

        // GET all posts
        var getPosts = Step.Create("GET /posts", async context =>
        {
            var request = Http.CreateRequest("GET", $"{BaseUrl}/posts");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // GET single post (random)
        var getPost = Step.Create("GET /posts/:id", async context =>
        {
            var postId = Random.Shared.Next(1, 101);
            var request = Http.CreateRequest("GET", $"{BaseUrl}/posts/{postId}");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // GET post comments
        var getComments = Step.Create("GET /posts/:id/comments", async context =>
        {
            var postId = Random.Shared.Next(1, 101);
            var request = Http.CreateRequest("GET", $"{BaseUrl}/posts/{postId}/comments");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // POST create post
        var createPost = Step.Create("POST /posts", async context =>
        {
            var body = $$$"""
                {
                    "title": "Load test post {{{Random.Shared.Next(1000)}}}",
                    "body": "Testing under load",
                    "userId": {{{Random.Shared.Next(1, 11)}}}
                }
                """;
            var request = Http.CreateRequest("POST", $"{BaseUrl}/posts")
                .WithHeader("Content-Type", "application/json")
                .WithBody(new StringContent(body));
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // PUT update post
        var updatePost = Step.Create("PUT /posts/:id", async context =>
        {
            var postId = Random.Shared.Next(1, 101);
            var body = $$$"""
                {
                    "id": {{{postId}}},
                    "title": "Updated title",
                    "body": "Updated body",
                    "userId": 1
                }
                """;
            var request = Http.CreateRequest("PUT", $"{BaseUrl}/posts/{postId}")
                .WithHeader("Content-Type", "application/json")
                .WithBody(new StringContent(body));
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // DELETE post
        var deletePost = Step.Create("DELETE /posts/:id", async context =>
        {
            var postId = Random.Shared.Next(1, 101);
            var request = Http.CreateRequest("DELETE", $"{BaseUrl}/posts/{postId}");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // GET users
        var getUsers = Step.Create("GET /users", async context =>
        {
            var request = Http.CreateRequest("GET", $"{BaseUrl}/users");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        // GET todos
        var getTodos = Step.Create("GET /todos", async context =>
        {
            var request = Http.CreateRequest("GET", $"{BaseUrl}/todos");
            var response = await Http.Send(httpClient, request);
            return response;
        });

        var scenario = ScenarioBuilder
            .CreateScenario("Load Test", getPosts, getPost, getComments, createPost, updatePost, deletePost, getUsers, getTodos)
            .WithWarmUpDuration(TimeSpan.FromSeconds(10))
            .WithLoadSimulations(
                Simulation.RampingInject(rate: 10, interval: TimeSpan.FromSeconds(1), during: TimeSpan.FromMinutes(2)),
                Simulation.Inject(rate: 20, interval: TimeSpan.FromSeconds(1), during: TimeSpan.FromMinutes(3)),
                Simulation.RampingInject(rate: 50, interval: TimeSpan.FromSeconds(1), during: TimeSpan.FromMinutes(2))
            );

        NBomberRunner
            .RegisterScenarios(scenario)
            .Run();
    }
}

File: locust/load_test.py
"""
Load Test - Comprehensive API testing with gradual ramp-up
Target: 10-50 users, 7 minutes
"""

from locust import task, between
from common import JSONPlaceholderUser


class LoadTestUser(JSONPlaceholderUser):
    """Load test user behavior"""

    wait_time = between(0.3, 0.5)  # Wait 0.3-0.5 seconds between tasks

    @task(5)
    def get_all_posts(self):
        """GET all posts"""
        self.client.get("/posts")

    @task(10)
    def get_single_post(self):
        """GET a random single post"""
        post_id = self.get_random_post_id()
        self.client.get(f"/posts/{post_id}")

    @task(5)
    def get_post_comments(self):
        """GET comments for a random post"""
        post_id = self.get_random_post_id()
        self.client.get(f"/posts/{post_id}/comments")

    @task(3)
    def create_post(self):
        """POST create a new post"""
        payload = {
            "title": f"Load test post {self.get_random_number()}",
            "body": "Testing under load",
            "userId": self.get_random_user_id()
        }
        self.client.post("/posts", json=payload)

    @task(2)
    def update_post(self):
        """PUT update a post"""
        post_id = self.get_random_post_id()
        payload = {
            "id": post_id,
            "title": "Updated title",
            "body": "Updated body",
            "userId": 1
        }
        self.client.put(f"/posts/{post_id}", json=payload)

    @task(2)
    def delete_post(self):
        """DELETE a post"""
        post_id = self.get_random_post_id()
        self.client.delete(f"/posts/{post_id}")

    @task(3)
    def get_users(self):
        """GET all users"""
        self.client.get("/users")

    @task(3)
    def get_todos(self):
        """GET all todos"""
        self.client.get("/todos")
